avalanche/training/supervised/der.py

```diff
@@ -0,0 +1,263 @@
+                            observed experiences (keys in replay_mem).
+        :param total_num_classes: If adaptive size is False, the fixed number
+                                  of classes to divide capacity over.
+        """
+        if not adaptive_size:
+            assert (
+                total_num_classes > 0
+            ), """When fixed exp mem size, total_num_classes should be > 0."""
+
+        super().__init__(max_size, adaptive_size, total_num_classes)
+        self.adaptive_size = adaptive_size
+        self.total_num_classes = total_num_classes
+        self.seen_classes = set()
+
+    def update(self, strategy: "SupervisedTemplate", **kwargs):
+        new_data = strategy.experience.dataset.eval()
+
+        new_data_with_logits = dataset_with_logits(
+            new_data, strategy.model, strategy.train_mb_size, strategy.device,
+        )
+
+        # Get sample idxs per class
+        cl_idxs = {}
+        for idx, target in enumerate(new_data.targets):
+            if target not in cl_idxs:
+                cl_idxs[target] = []
+            cl_idxs[target].append(idx)
+
+        # Make AvalancheSubset per class
+        cl_datasets = {}
+        for c, c_idxs in cl_idxs.items():
+            subset = classification_subset(new_data_with_logits, indices=c_idxs)
+            cl_datasets[c] = subset
+        # Update seen classes
+        self.seen_classes.update(cl_datasets.keys())
+
+        # associate lengths to classes
+        lens = self.get_group_lengths(len(self.seen_classes))
+        class_to_len = {}
+        for class_id, ll in zip(self.seen_classes, lens):
+            class_to_len[class_id] = ll
+
+        # update buffers with new data
+        for class_id, new_data_c in cl_datasets.items():
+            ll = class_to_len[class_id]
+            if class_id in self.buffer_groups:
+                old_buffer_c = self.buffer_groups[class_id]
+                # Here it uses underlying dataset
+                old_buffer_c.update_from_dataset(new_data_c)
+                old_buffer_c.resize(strategy, ll)
+            else:
+                new_buffer = ReservoirSamplingBuffer(ll)
+                new_buffer.update_from_dataset(new_data_c)
+                self.buffer_groups[class_id] = new_buffer
+
+        # resize buffers
+        for class_id, class_buf in self.buffer_groups.items():
+            self.buffer_groups[class_id].resize(
+                strategy, 
+                class_to_len[class_id])
+
+
+class DER(SupervisedTemplate):
+    """ 
+    Implements the DER and the DER++ Strategy, 
+    from the "Dark Experience For General Continual Learning" 
+    paper, Buzzega et. al, https://arxiv.org/abs/2004.07211 
+    """
+
+    def __init__(
+        self,
+        model: Module,
+        optimizer: Optimizer,
+        criterion=CrossEntropyLoss(),
+        mem_size: int = 200,
+        batch_size_mem: int = None,
+        derpp: bool = False,
```

Comment:
I would remove this argument and suggest the user to set `beta=0`.
Otherwise, if a user sets `beta` but forgets `derpp` it will silently ignore the argument.